gh-74756: support precision field for integer formatting types

skirpichev opened the pull request:
For integer presentation types (excluding ``'c'``), the precision gives the minimal number of digits to appear, expanded with an appropriate number of leading zeros.

If ``'z'`` option specified for non-decimal presentation types - integer value interpreted as two's complement, the precision gives it's minimum size ``precision*k`` in bits, where ``k=1,3,4`` for ``'b'``, ``'o'`` and ``'x'``/``'X'`` types, respectively.

A precision of ``0`` is treated as equivalent to a precision of ``1``.

Examples:
```pycon
>>> f"{-12:z.8b}"
'11110100'
>>> f"{-12:#.8b}"
'-0b00001100'
>>> f"{200:z.8b}"
'011001000'
>>> f"{200:.8b}"
'11001000'
>>> f"{123:.8d}"
'00000123'
>>> f"{-12:.8d}"
'-00000012'
>>> f"{-129:z#.2x}"
'0xf7f'
>>> f"{-129:z#.3x}"
'0xf7f'
>>> f"{-129:z#.4x}"
'0xff7f'
>>> f"{383 :z#.2x}"
'0x17f'
>>> f"{383 :z#.3x}"
'0x17f'
>>> f"{383 :z#.4x}"
'0x017f'
```

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-74756 -->
* Issue: gh-74756
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--131926.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

skirpichev committed:
gh-74756: support precision field for integer formatting types

```pycon
>>> f"{-12:.8b}"
'11110100'
>>> f"{200:.8b}"
Traceback (most recent call last):
  File "<python-input-5>", line 1, in <module>
    f"{200:.8b}"
      ^^^^^^^^^
OverflowError: Expected integer in range [-2**7, 2**7)
>>> f"{123:.8d}"
'00000123'
>>> f"{-12:.8d}"
'-00000012'
```

skirpichev committed:
address review: exceptions

skirpichev committed:
+ cleanup

jb2170 commented on Python/formatter_unicode.c:
```diff
@@ -979,12 +980,6 @@ format_long_internal(PyObject *value, const InternalFormatSpec *format,
        from a hard-code pseudo-locale */
     LocaleInfo locale = LocaleInfo_STATIC_INIT;
 
-    /* no precision allowed on integers */
```
Should this be kept for the `c` character type? Currently one can do:
```py
f"{65:.5c}" # 'A'
f"{3486:.100c}" # 'ඞ'
```
which silently skips the precision. If this is unintended perhaps it should be moved into the `/* special case for character formatting */`?

skirpichev committed:
Merge branch 'master' into format-twos-compl/74756

skirpichev committed:
address review: disallow precision for 'c' type

skirpichev committed:
+1 // docs

skirpichev committed:
Merge branch 'master' into format-twos-compl/74756

skirpichev committed:
address review: don't raise ValueError (precision specifies the *minimal* range)

skirpichev committed:
address review: use z option for twos complement interpretation

skirpichev committed:
+ cleanup and docs

skirpichev committed:
Merge branch 'master' into format-twos-compl/74756

skirpichev committed:
address review: correct precision in two's complement case

skirpichev committed:
Merge branch 'master' into format-twos-compl/74756

skirpichev committed:
address review: ensure leading bit is 1 for negatives

now we choose minimal twos complement size as being >= k*precision
AND with leading bit set to 1 for negatives

Increasing precision will adds 0's of (base-1)'s as needed:

>>> f"{-129:z#.2x}"
'0xf7f'
>>> f"{-129:z#.3x}"
'0xf7f'
>>> f"{-129:z#.4x}"
'0xff7f'
>>> f"{383 :z#.2x}"
'0x17f'
>>> f"{383 :z#.3x}"
'0x17f'
>>> f"{383 :z#.4x}"
'0x017f'

skirpichev committed:
Merge branch 'master' into format-twos-compl/74756

skirpichev committed:
+ check if nbits is too big (fix warnings)
